internal/expr/parser.go

```diff
@@ -265,28 +272,174 @@ func (p *Parser) parseIdentifier() (string, bool) {
+		}
+
+		idField, ok := p.parseIdentifierAsterisk()
+		if !ok {
+			return fullName{}, false, fmt.Errorf("invalid identifier near char %d", p.pos)
 		}
-		return fullName{}, false,
-			fmt.Errorf("go object near char %d not qualified", p.pos)
+		return fullName{id, idField}, true, nil
 	}
+
 	cp.restore()
 	return fullName{}, false, nil
 }
 
-// parseInputExpression parses an input expression of the form $Type.name.
+// parseList takes a parsing function that returns a fullName and parses a
+// bracketed, comma seperated, list.
+func (p *Parser) parseList(parseFn func(p *Parser) (fullName, bool, error)) ([]fullName, bool, error) {
+	cp := p.save()
+
+	if !p.skipByte('(') {
+		return nil, false, nil
+	}
+
+	parenPos := p.pos
+	p.skipSpaces()
+
+	nextItem := true
+	var objs []fullName
+	for i := 0; nextItem; i++ {
+		if obj, ok, err := parseFn(p); ok {
+			objs = append(objs, obj)
+		} else if err != nil {
+			return nil, false, err
+		} else if i == 0 {
+			// If the first item is not what we are looking for, we exit.
+			cp.restore()
+			return nil, false, nil
+		} else {
+			// On subsequent items we return an error.
+			return nil, false, fmt.Errorf("column %d: invalid expression", p.pos)
+		}
+
+		p.skipSpaces()
+		if p.skipByte(')') {
+			return objs, true, nil
+		}
+
+		nextItem = p.skipByte(',')
+		p.skipSpaces()
+	}
+	return nil, false, fmt.Errorf("column %d: missing closing parentheses", parenPos)
+}
+
+// parseColumns parses a list of columns. For lists of more than one column the
+// columns must be enclosed in brackets e.g. "(col1, col2) AS &Person.*".
+func (p *Parser) parseColumns() ([]fullName, bool) {
+	cp := p.save()
+
+	// Case 1: A single column e.g. "p.name".
+	if col, ok, _ := p.parseColumn(); ok {
+		return []fullName{col}, true
+	}
+
+	// Case 2: Multiple columns e.g. "(p.name, p.id)".
+	if cols, ok, _ := p.parseList((*Parser).parseColumn); ok {
+		return cols, true
+	}
+
+	cp.restore()
+	return nil, false
+}
+
+// parseTargets parses the part of the output expression following the
+// ampersand. This can be one or more references Go objects.
+// If the ampersand is not preceded by a space and succeeded by a name or
+// opening bracket the targets are not parsed.
+func (p *Parser) parseTargets() ([]fullName, bool, error) {
+	cp := p.save()
+
+	// Case 1: A single target e.g. "&Person.name".
+	if target, ok, err := p.parseTarget(); err != nil {
+		return nil, false, err
+	} else if ok {
+		return []fullName{target}, true, nil
+	}
+
+	// Case 2: Multiple targets e.g. "(&Person.name, &Person.id)".
+	if targets, ok, err := p.parseList((*Parser).parseTarget); err != nil {
+		return nil, false, err
+	} else if ok {
+		return targets, true, nil
+	}
+
+	cp.restore()
+	return nil, false, nil
+}
+
+// parseOutputExpression parses all output expressions. The ampersand must be
+// preceded by a space and followed by a name byte.
+func (p *Parser) parseOutputExpression() (*outputPart, bool, error) {
+	cp := p.save()
```

Comment:
Similar here, but here we actually need the save/restore before the parseColumsn branch, since we're moving the pointer forward there and may still fail for other reasons.

Reply:
Changed.